### PR TITLE
Topic black scholes

### PR DESCRIPTION
**Pull request by vincentridder**
Ported example BlackScholas from cuda sdk. And some macros to map needed Cuda math functions to alpaka

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
Please squash your commits that is become more easy to follow your changes

File: example/CUDASamples/BlackScholes/src/BlackScholes_kernel.cuh [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            float Volatility,
+            int optN
+    ) const
+    {
+        ////Thread index
+        //const int      tid = blockDim.x * blockIdx.x + threadIdx.x;
+        ////Total number of threads in execution grid
+        //const int THREAD_N = blockDim.x * gridDim.x;
+
+        const int opt_begin = blockDim.x * blockIdx.x *elemDim.x + threadIdx.x;
```
Comment (psychocoderHPC) [inline example/CUDASamples/BlackScholes/src/BlackScholes_kernel.cuh]: Should be `const int opt_begin = blockDim.x * blockIdx.x *elemDim.x + threadIdx.x * elemDim.x;`

File: include/cupla/cudaToCupla/runtime.hpp [hunk 2]
```diff
@@ -77,3 +77,11 @@
 #define make_cudaPos(...) make_cuplaPos(__VA_ARGS__)
 
 #define make_cudaPitchedPtr(...) make_cuplaPitchedPtr(__VA_ARGS__)
+//math
+#define __fdividef(a,b) ((a)/(b))
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/runtime.hpp]: please create a own pull request for this changes because they are independent of the example and extent the core part of **cupla**

File: example/CUDASamples/BlackScholes/src/BlackScholes_kernel.cuh [hunk 3]
```diff
@@ -105,7 +105,7 @@ struct BlackScholesGPU {
         ////Total number of threads in execution grid
         //const int THREAD_N = blockDim.x * gridDim.x;
 
-        const int opt_begin = blockDim.x * blockIdx.x *elemDim.x + threadIdx.x;
+        const int opt_begin = blockDim.x * blockIdx.x *elemDim.x + threadIdx.x* elemDim.x;
```
Comment (psychocoderHPC) [inline example/CUDASamples/BlackScholes/src/BlackScholes_kernel.cuh]: space missing befor `* elemDim.x`

File: include/cupla/cudaToCupla/runtime.hpp [hunk 4]
```diff
@@ -77,3 +77,10 @@
 #define make_cudaPos(...) make_cuplaPos(__VA_ARGS__)
 
 #define make_cudaPitchedPtr(...) make_cuplaPitchedPtr(__VA_ARGS__)
+
+//math
+#define __fdividef(a,b) ((a)/(b))
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/runtime.hpp]: please move this changes to a separate pull request, after I merged the pull you can rebase this pull request.

File: include/cupla/cudaToCupla/runtime.hpp [hunk 5]
```diff
@@ -77,3 +77,10 @@
...
+#define fabsf(a) alpaka::math::abs(acc,a)
+#define __expf(a) alpaka::math::exp(acc,a)
+#define __logf(a) alpaka::math::log(acc,a)
+#define rsqrtf(a) alpaka::math::rsqrt(acc,a)
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/runtime.hpp]: btw. empty line at the end of file is missing

File: example/CUDASamples/BlackScholes/src/BlackScholes_kernel.cuh [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            float Volatility,
+            int optN
+    ) const
+    {
+        ////Thread index
+        //const int      tid = blockDim.x * blockIdx.x + threadIdx.x;
+        ////Total number of threads in execution grid
+        //const int THREAD_N = blockDim.x * gridDim.x;
+
+        const int opt_begin = blockDim.x * blockIdx.x *elemDim.x + threadIdx.x* elemDim.x;
```
Comment (psychocoderHPC) [inline example/CUDASamples/BlackScholes/src/BlackScholes_kernel.cuh]: space before `elemDim` missing

**Comment by psychocoderHPC** [comment]
thanks for this new example. One small change is needed.

**Comment by psychocoderHPC** [comment]
The clang source on http://llvm.org/apt/ are currently down.  Therefore the tests fails :-(

File: include/cupla/cudaToCupla/runtime.hpp [hunk 7]
```diff
@@ -84,3 +84,5 @@
 #define __expf(a) alpaka::math::exp(acc,a)
 #define __logf(a) alpaka::math::log(acc,a)
 #define rsqrtf(a) alpaka::math::rsqrt(acc,a)
+
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/runtime.hpp]: please remove this changes

**Comment by psychocoderHPC** [comment]
please remove the marked line and rebase to the newest dev

**Comment by sbastrakov** [comment]
@psychocoderHPC as far as i can see, this example is nearly done, and could be useful to demonstrate cupla `__device__` and equivalents of CUDA fast math functions. Shall we finish it?

**Comment by psychocoderHPC** [comment]
> @psychocoderHPC as far as i can see, this example is nearly done, and could be useful to demonstrate cupla `__device__` and equivalents of CUDA fast math functions. Shall we finish it?

Feel free to finish it but only if it is not costs more than 2h

File: example/CUDASamples/BlackScholes/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+# along with cupla.
+# If not, see <http://www.gnu.org/licenses/>.
+#
+
+
+################################################################################
+# Required CMake version.
+################################################################################
+
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8.12)
```
Comment (ax3l) [inline example/CUDASamples/BlackScholes/CMakeLists.txt]: please use lower-case cmake commands, as we try to use in the main cmake files in the repo.

**Comment by sbastrakov** [comment]
This functionality is being added as #153, closing this PR.
